File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0 
        
        # maxDepth must occur at root, so we don't need to traverse check like 543 Diameter of Binary Tree
                
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right)) 
        
        
        
        

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root)
            return 0;
        int depth = 0;
        queue<TreeNode*> bfs;
        bfs.push(root);
        while (!bfs.empty()) {
            int curNodeN = bfs.size();
            for (int i = 0; i < curNodeN; i++) {
                TreeNode* node = bfs.front();
                bfs.pop();
                if (node->left)
                    bfs.push(node->left);
                if (node->right)
                    bfs.push(node->right);
            }
            depth++;            
        }
        return depth;        
    }
};